qiskit_experiments/data_processing/nodes.py

```diff
@@ -596,3 +599,170 @@ class ProjectorType(Enum):
+
+    def _reorder(self, unordered_data: np.array) -> np.array:
+        """Reorder the measured data according to the measurement sequence.
+
+        Here, is assumed that the inner loop of the measurement is done over the circuits
+        and the outer loop is done over the shots.
+        """
+        if unordered_data is None:
+            return unordered_data
+
+        order_data = []
+
+        for shot_idx in range(self._n_shots):
+            for circuit_idx in range(self._n_circuits):
+                order_data.append(unordered_data[circuit_idx][shot_idx])
+
+        return np.array(order_data)
+
+
+class RestlessToCounts(RestlessNode):
+    """Convert restless memory to counts.
+
+    This node takes as input a list of lists where the sublist is the memory of
+    each measured circuit. The sublists therefore have a length given by the
+    number of shots. This data is reordered into a one dimensional array where
+    the element at index j was the jth measured shot. This node assumes that
+    a list of circuits :code:`[circ_1, cric_2, ..., circ_m]` is measured :code:`n_shots`
+    times according to the following order:
+
+    .. parsed-literal::
+
+        [
+            circuit 1 - shot 1,
+            circuit 2 - shot 1,
+            ...
+            circuit m - shot 1,
+            circuit 1 - shot 2,
+            circuit 2 - shot 2,
+            ...
+            circuit m - shot 2,
+            circuit 1 - shot 3,
+            ...
+            circuit m - shot n,
+        ]
+
+    Once the shots have been ordered in this fashion the node compares each shot with the
+    previous shot. If they are the same then the shot corresponds to a 0, i.e. no state
+    change, and if they are different then the shot corresponds to a 1, i.e. there was
+    a state change.
+    """
+
+    def __init__(self, num_qubits: int, validate: bool = True):
+        """
+        Args:
+            num_qubits: The number of qubits which is needed to construct the header needed
+            by :code:`qiskit.result.postprocess.format_counts_memory` to convert the memory
+            into a bit-string of counts.
+            validate: If set to False the DataAction will not validate its input.
+        """
+        super().__init__(validate)
+        self._num_qubits = num_qubits
+
+    def _process(self, data: np.array) -> np.array:
+        """Reorder the shots and assign values to them based on the previous outcome.
+
+        Args:
+            data: An array representing the memory.
+
+        Returns:
+            A counts dictionary processed according to the restless methodology.
+        """
+
+        # Step 1. Reorder the data.
+        memory = self._reorder(data)
+
+        # Step 2. Do the restless classification into counts.
+        counts = [defaultdict(int) for _ in range(self._n_circuits)]
+        prev_shot = "0" * self._num_qubits
+        header = {"memory_slots": self._num_qubits}
+
+        for idx, shot in enumerate(memory):
```

Comment:
You may want to do something like
```python
data = np.array(["0000", "0010", "0010", "0000", "0000", "0001"])
diff_loc = np.where(data[1:] != data[:-1])

new_data = np.zeros(data.shape)
new_data[diff_loc] = 1  # array([1., 0., 1., 0., 1., 0.])
```
since python has type inference overhead, looping over every element should be avoided as much as possible. Array operations will give you significant performance improvement.

Another question: is there any situation we may want to consider measurement error? Or this protocol is completely error-free?

Reply:
I see the idea behind this suggestion but the code does not do restless processing. Indeed, we do not want to compare the entire string but XOR each bit in the string with its position in the previous string. Consider the two measurement outcomes `"01"` and `"11"`. Here, `"01"` came first. These two bit strings are not identical so `"01" != "11"` is true. but the bit string we want to get out of the data processing is `"10"` since qubit 1 changed state from one measurement to the next. We could probably achieve this by putting the data into a two-dimensional array with time on one axis and qubit index on the other axis to do the diff but the cost of bringing the data into this format will probably outweigh any benefit. I would therefore leave as is.

Doing measurement error mitigation could perhaps be done on the reclassified count dictionaries. I'd leave that to another PR.